# Schema Validation — Olist E-Commerce Dataset

This notebook validates the structure and key constraints of the Olist dataset before setting up the relational database.

The main focus is on:
- identifying candidate primary keys
- validating uniqueness
- checking null values
- understanding identifiers and relationships

## 1. Customers

### 1.1 Primary Key Validation — `customer_id`

#### Question

Can `customer_id` serve as the primary key of the `customers` table?

To answer this, we check:
- total number of records
- number of unique `customer_id` values
- number of null `customer_id` values

In [2]:
import pandas as pd

customers_df = pd.read_csv("../data/olist_customers_dataset.csv")

customers_df.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


In [3]:
customers_df.shape[0]

99441

In [4]:
customers_df["customer_id"].nunique()

99441

In [5]:
customers_df["customer_id"].isna().sum()

np.int64(0)

#### Insight

`customer_id` contains 99,441 unique and non-null values across 99,441 records. Therefore, it satisfies the uniqueness and non-null requirements for a primary key in the `customers` table.

### 1.2 Customer Identifier Validation — `customer_unique_id`

#### Question

Is `customer_unique_id` unique for each record in the `customers` table, or can the same customer appear in multiple records?

To answer this, we compare:
- total number of records
- number of unique `customer_unique_id` values
- number of null `customer_unique_id` values

In [6]:
customers_df.shape[0]

99441

In [7]:
customers_df["customer_unique_id"].nunique()

96096

In [8]:
customers_df["customer_unique_id"].isna().sum()

np.int64(0)

#### Insight

`customer_id` is unique for every record, while `customer_unique_id` is not unique: there are 99,441 records but only 96,096 unique `customer_unique_id` values.

This suggests that `customer_id` identifies a customer record, whereas `customer_unique_id` represents a customer at the business level and may appear in multiple records.

The relationship between repeated `customer_unique_id` values and multiple orders will be validated later using the `orders` table.

## 2. Orders

In [9]:
orders_df = pd.read_csv("../data/olist_orders_dataset.csv")

### 2.1 Primary Key Validation — `order_id`

#### Question

Can `order_id` serve as the primary key of the `orders` table?

To answer this, we check:
- total number of records
- number of unique `order_id` values
- number of null `order_id` values

In [10]:
orders_df.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


In [11]:
orders_df.shape[0]

99441

In [12]:
orders_df["order_id"].nunique()

99441

In [13]:
orders_df["order_id"].isna().sum()

np.int64(0)

### Insight

`order_id` contains 99,441 unique and non-null values across 99,441
records. Therefore, it satisfies the uniqueness and non-null requirements
for a primary key in the `orders` table.

### 2.2 Foreign Key Validation — `customer_id`

#### Question

Does `orders.customer_id` correctly reference `customers.customer_id`?

To answer this, we need to check:
- whether `orders.customer_id` contains null values
- whether every `customer_id` in `orders` exists in the `customers` table

In [14]:
orders_df["customer_id"].isna().sum()

np.int64(0)

In [15]:
(~orders_df["customer_id"].isin(customers_df["customer_id"])).sum()

np.int64(0)

#### Insight

All `customer_id` values in `orders` exist in `customers`, so `customer_id` is a valid foreign key.

## 3. Order Items

In [16]:
order_items_df = pd.read_csv("../data/olist_order_items_dataset.csv")

In [17]:
order_items_df.head(10)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
5,00048cc3ae777c65dbb7d2a0634bc1ea,1,ef92defde845ab8450f9d70c526ef70f,6426d21aca402a131fc0a5d0960a3c90,2017-05-23 03:55:27,21.90,12.69
6,00054e8431b9d7675808bcb819fb4a32,1,8d4f2bb7e93e6710a28f34fa83ee7d28,7040e82f899a04d1b434b795a43b4617,2017-12-14 12:10:31,19.90,11.85
7,000576fe39319847cbb9d288c5617fa6,1,557d850972a7d6f792fd18ae1400d9b6,5996cddab893a4652a15592fb58ab8db,2018-07-10 12:30:45,810.00,70.75
8,0005a1a1728c9d785b8e2b08b904576c,1,310ae3c140ff94b03219ad0adc3c778f,a416b6a846a11724393025641d4edd5e,2018-03-26 18:31:29,145.95,11.65
9,0005f50442cb953dcd1d21e1fb923495,1,4535b0e1091c278dfd193e5a1d63b39f,ba143b05f0110f0dc71ad71b4466ce92,2018-07-06 14:10:56,53.99,11.40


### 3.1 Primary Key Validation - `order_id`

#### Question

Can `order_id` alone serve as the primary key of the `order_items` table?

To answer this, we compare:

- total number of records
- number of unique `order_items.order_id` values
- number of null `order_items.order_id` values

In [18]:
order_items_df.shape[0]

112650

In [19]:
order_items_df["order_id"].nunique()

98666

In [20]:
order_items_df["order_id"].isna().sum()

np.int64(0)

So `order_id` alone cannot be the primary key, because multiple rows belong to the same order.

### 3.2 Composite Primary Key Validation — `order_id`, `order_item_id`

#### Question

Can the combination of `order_id` and `order_item_id` uniquely identify each row in the `order_items` table?

To answer this, we need to check whether each combination of `order_id` and `order_item_id` is unique and non-null.

In [21]:
order_items_df[["order_id", "order_item_id"]].shape[0]

112650

In [22]:
order_items_df[["order_id", "order_item_id"]].drop_duplicates().shape[0]

112650

In [23]:
order_items_df[["order_id", "order_item_id"]].isna().sum()

order_id         0
order_item_id    0
dtype: int64

#### Insight

The combination of `order_id` and `order_item_id` is unique and non-null across all 112,650 records. Therefore, `(order_id, order_item_id)` can serve as the composite primary key of `order_items`.

### 3.3 Foreign Key Validation - `order_id`

#### Question 
Does every `order_id` in `order_items` exist in the `orders` table?

In [24]:
(~order_items_df["order_id"].isin(orders_df["order_id"])).sum()

np.int64(0)

In [25]:
order_items_df["order_id"].isna().sum()

np.int64(0)

All `order_id` values in `order_items` exist in the `orders` table, and none are null. Therefore, `order_items.order_id` satisfies the requirements for a foreign key.

### 3.4 Foreign Key Validation - `product_id`

#### Question 
Does every `product_id` in `order_items` exist in the `products` table?

we first load the products table so that its primary key can be used for referential integrity checks.

In [26]:
products_df = pd.read_csv("../data/olist_products_dataset.csv")

In [27]:
products_df.shape[0]

32951

In [28]:
products_df.head(10)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0
5,41d3672d4792049fa1779bb35283ed13,instrumentos_musicais,60.0,745.0,1.0,200.0,38.0,5.0,11.0
6,732bd381ad09e530fe0a5f457d81becb,cool_stuff,56.0,1272.0,4.0,18350.0,70.0,24.0,44.0
7,2548af3e6e77a690cf3eb6368e9ab61e,moveis_decoracao,56.0,184.0,2.0,900.0,40.0,8.0,40.0
8,37cc742be07708b53a98702e77a21a02,eletrodomesticos,57.0,163.0,1.0,400.0,27.0,13.0,17.0
9,8c92109888e8cdf9d66dc7e463025574,brinquedos,36.0,1156.0,1.0,600.0,17.0,10.0,12.0


In [29]:
order_items_df["product_id"].isna().sum()

np.int64(0)

In [30]:
(~order_items_df["product_id"].isin(products_df["product_id"])).sum()

np.int64(0)

#### Insight

All `product_id` values in `order_items` exist in the `products` table, and none are null. Therefore, `order_items.product_id` is a valid foreign key.

### 3.5 Foreign Key Validation - `seller_id`

#### Question 
Does every `seller_id` in `order_items` exist in the `sellers` table?

Before the validation, we need `seller_df`. Since it hasn't been loaded yet, first load:

In [31]:
sellers_df = pd.read_csv("../data/olist_sellers_dataset.csv")

In [32]:
sellers_df["seller_id"].isna().sum()

np.int64(0)

In [33]:
(~order_items_df["seller_id"].isin(sellers_df["seller_id"])).sum()

np.int64(0)

#### Insight

All `seller_id` values in `order_items` exist in the `sellers` table, and none are null. Therefore, `order_items.seller_id` is a valid foreign key.

## 4. Order Payments

In [34]:
order_payments_df = pd.read_csv("../data/olist_order_payments_dataset.csv")

order_payments_df.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


### 4.1 Primary Key Validation - `order_id`

#### Question:
Can `order_id` serve as the primary key of `order_payments`?

In [35]:
order_payments_df.shape[0]

103886

In [36]:
order_payments_df["order_id"].nunique()

99440

In [37]:
order_payments_df["order_id"].isna().sum()

np.int64(0)

#### Insight

So `order_id` alone cannot be the primary key, because multiple rows belong to the same order payment.

### 4.2 Candidate Composite Primary Key - `order_id` + `payment_sequential`

#### Question
Can the combination `(order_id, payment_sequential)` uniquely identify each payment record?

In [38]:
order_payments_df.shape[0]

103886

In [39]:
order_payments_df[["order_id", "payment_sequential"]].drop_duplicates().shape[0]

103886

In [40]:
order_payments_df[["order_id", "payment_sequential"]].isna().sum()

order_id              0
payment_sequential    0
dtype: int64

#### Insight

The combination of `order_id` and `payment_sequential` is unique and non-null across all 103886 records. Therefore, it can serve as the composite primary key of `order_payments`.

### 4.3 Foreign Key Validation - `order_id`

#### Question 
Does every `order_id` in`order_payments` exist in the `orders` table?

In [41]:
order_payments_df["order_id"].isna().sum()

np.int64(0)

In [42]:
(~order_payments_df["order_id"].isin(orders_df["order_id"])).sum()

np.int64(0)

#### Insight 
All `order_id` values in `order_payments` exist in the `orders` table, with no null or orphan references. Therefore, `order_id` is a valid foreign key to orders.

### 4.4 Identify Remaining Relationships

#### Question
Are there any additional foreign key relationships in `order_payments` that need to be validated?

#### Insight
No additional foreign key relationships were identified in `order_payments`. The remaining columns describe payment attributes rather than references to other tables.

## 5. Order Reviews

In [43]:
order_reviews_df = pd.read_csv("../data/olist_order_reviews_dataset.csv")

order_reviews_df.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


### 5.1 Primary Key Validation - `review_id` 

#### Question 
Can `review_id` uniquely identify each record?

In [44]:
order_reviews_df.shape[0]

99224

In [45]:
order_reviews_df["review_id"].nunique()

98410

In [46]:
order_reviews_df["review_id"].isna().sum()

np.int64(0)

#### Insight
`review_id` is non-null but not unique across all records. Therefore, `review_id` alone cannot serve as the primary key.

### 5.2 Composite Primary Key Validation - `review_id` + `order_id`

#### Question
Can the combination of `review_id` and `order_id` uniquely identify each review record?

In [47]:
order_reviews_df.shape[0]

99224

In [48]:
order_reviews_df[["review_id", "order_id"]].drop_duplicates().shape[0]

99224

In [49]:
order_reviews_df[["review_id", "order_id"]].isna().sum()

review_id    0
order_id     0
dtype: int64

#### Insight
The combination of `review_id` and `order_id` is unique and non-null across all 99,224 records. Therefore, it can serve as the composite primary key of `order_reviews`.

### 5.3 Foreign Key Validation - `order_id`

#### Question
Does every `order_id` in `order_reviews` exist in the `orders` table?

In [50]:
order_reviews_df["order_id"].isna().sum()

np.int64(0)

In [51]:
(~order_reviews_df["order_id"].isin(orders_df["order_id"])).sum()

np.int64(0)

#### Insight 
All `order_id` values in `order_reviews` exist in the `orders` table, with no null or orphan references. Therefore, `order_id` is a valid foreign key to `orders`.

### 5.4 Foreign Key Review 

#### Question 
Are there any additional foriegn key relationships in `order_reviews` that need to be validated?

#### Insight
No additional foreign key relationships were identified in `order_reviews`. The remaining columns represent review identifiers, scores, comments, and timestamps rather than references to other tables.

## 6. Products
The `products` table was already loaded in Section 3.4 for foreign key validation.

### 6.1 Primary Key Validation - `product_id`

#### Question 
Can `product_id` uniquely identify every product record?

In [52]:
products_df.shape[0]

32951

In [53]:
products_df["product_id"].nunique()

32951

In [54]:
products_df["product_id"].isna().sum()

np.int64(0)

#### Insight

`product_id` is unique and non-null across all 32,951 product records. Therefore, it can serve as the primary key of `products`.

### 6.2 Foreign Key Validation - `product_category_name`

#### Question

Does every non-null `product_category_name` in `products` have a corresponding entry in `product_category_name_translation`?

(Does every category used by Products exist in Translation?)

- First, we need to load `product_category_name_translation` into a DataFrame.

In [55]:
product_category_name_translation_df = pd.read_csv("../data/product_category_name_translation.csv")

product_category_name_translation_df.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [56]:
products_df["product_category_name"].isna().sum()

np.int64(610)

#### Insight

There are 610 products with a missing `product_category_name`. Therefore, this relationship is not mandatory for every product record. However, null values alone do not disqualify `product_category_name` from being a foreign key. We still need to verify whether all non-null category values exist in the translation table.


In [57]:
(~products_df[~products_df["product_category_name"].isna()]["product_category_name"].isin(product_category_name_translation_df["product_category_name"])).sum()

np.int64(13)

#### Insight

Among the non-null `product_category_name` values, 13 categories have no corresponding entry in the translation table. Therefore, the relationship does not currently satisfy referential integrity and cannot be implemented as a foreign key without addressing these unmatched values.

### 6.3 Candidate Key Validation — `product_category_name`

#### Question

Can `product_category_name` uniquely identify each row in `product_category_name_translation`?

(If I find "books" in Translation, does i point to exactly ONE row?)

In [58]:
product_category_name_translation_df.shape[0]

71

In [59]:
product_category_name_translation_df["product_category_name"].nunique()

71

In [60]:
product_category_name_translation_df["product_category_name"].isna().sum()

np.int64(0)

#### Insight

`product_category_name` is unique and non-null across all records in the translation table. Therefore, it can serve as a candidate key for `product_category_name_translation` and can be referenced by a foreign key.

## 7. Sellers

The `Sellers` table was already loaded in previous sections.

### 7.1 Primary Key Validation - `seller_id`

#### Question 
Can `seller_id` uniquely identify each seller record?

In [61]:
sellers_df.shape[0]

3095

In [62]:
sellers_df["seller_id"].nunique()

3095

In [63]:
sellers_df["seller_id"].isna().sum()

np.int64(0)

#### Insight

`seller_id` is unique and non-null across all 3095 seller records. Therefore, it can serve as the primary key of `sellers`.

### 7.2 Foreign Key Review

#### Question

Are there any additional foreign key relationships in `sellers` that need to be validated?

#### Insight

No additional foreign key relationships were identified in `sellers`. The remaining columns describe the seller's location and do not represent references to other tables.

## 8. Geolocation

In [64]:
geolocation_df = pd.read_csv("../data/olist_geolocation_dataset.csv")

geolocation_df.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


### 8.1 Primary Key Validation - `geolocation_zip_code_prefix`

#### Question
Can `geolocation_zip_code_prefix` uniquely identify each geolocation record?

In [65]:
geolocation_df.shape[0]

1000163

In [66]:
geolocation_df["geolocation_zip_code_prefix"].nunique()

19015

In [67]:
geolocation_df["geolocation_zip_code_prefix"].isna().sum()

np.int64(0)


#### Insight

`geolocation_zip_code_prefix` is non-null but not unique across the 1,000,163 records. Therefore, it cannot serve as the primary key. Multiple records can exist for the same ZIP-code prefix, so the table contains multiple geographic observations per prefix.

### 8.2 Candidate Composite Key — `geolocation_zip_code_prefix` + `geolocation_lat`

#### Question

Can the combination of `geolocation_zip_code_prefix` and `geolocation_lat` uniquely identify each geolocation record?


In [68]:
geolocation_df.shape[0]

1000163

In [69]:
geolocation_df[["geolocation_zip_code_prefix", "geolocation_lat"]].drop_duplicates().shape[0]

719300

#### Insight

The combination is not unique: there are 719,300 unique pairs across 1,000,163 records. Therefore, `geolocation_zip_code_prefix` and `geolocation_lat` together cannot serve as a primary key.

### 8.3 Candidate Composite Key — `geolocation_zip_code_prefix` + `geolocation_lat` + `geolocation_lng`

#### Question

Can the combination of `geolocation_zip_code_prefix`, `geolocation_lat`, and `geolocation_lng` uniquely identify each geolocation record?

In [70]:
geolocation_df.shape[0]

1000163

In [71]:
geolocation_df[["geolocation_zip_code_prefix", "geolocation_lat", "geolocation_lng"]].drop_duplicates().shape[0]

720154

#### Insight

The combination is not unique: there are 720,154 unique triples across 1,000,163 records. Therefore, `geolocation_zip_code_prefix` + `geolocation_lat` + `geolocation_lng` cannot serve as a primary key.

### 8.4 Candidate Composite Key — `geolocation_lat` + `geolocation_lng`

#### Question

Can the combination of `geolocation_lat` and `geolocation_lng` uniquely identify each geolocation record?

In [72]:
geolocation_df.shape[0]

1000163

In [73]:
geolocation_df[["geolocation_lat", "geolocation_lng"]].drop_duplicates().shape[0]

718463

#### Insight

The combination of `geolocation_lat` and `geolocation_lng` is not unique across all records. Therefore, it cannot serve as a primary key for the geolocation table.

### 8.5 Investigating Duplicate Geographic Records

#### Question

Do multiple records contain exactly the same geographic information?

In [74]:
geolocation_df[
    ["geolocation_zip_code_prefix",
     "geolocation_lat",
     "geolocation_lng"]
].value_counts().head(10)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
88220                        -27.102099       -48.629613         314
6414                         -23.495901       -46.874687         189
5145                         -23.506049       -46.717377         141
6414                         -23.490618       -46.869004         127
22620                        -23.005514       -43.375964         102
22640                        -23.004582       -43.319899          89
22775                        -22.965906       -43.389999          89
71936                        -15.841451       -48.024026          85
3015                         -23.537186       -46.594036          83
9781                         -23.727641       -46.531065          81
Name: count, dtype: int64

In [75]:
geolocation_df[
    ["geolocation_zip_code_prefix",
     "geolocation_lat",
     "geolocation_lng",
     "geolocation_city",
     "geolocation_state"]
].value_counts().head(10)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  geolocation_city  geolocation_state
88220                        -27.102099       -48.629613       itapema           SC                   314
6414                         -23.495901       -46.874687       barueri           SP                   189
                             -23.490618       -46.869004       barueri           SP                   127
5145                         -23.506049       -46.717377       sao paulo         SP                   126
22620                        -23.005514       -43.375964       rio de janeiro    RJ                   102
22775                        -22.965906       -43.389999       rio de janeiro    RJ                    89
22640                        -23.004582       -43.319899       rio de janeiro    RJ                    89
6401                         -23.509243       -46.886669       barueri           SP                    81
71936                        -15.841451       -48.024

#### Insight

Some complete geographic records are repeated multiple times in the dataset. Therefore, the combination of the available geographic attributes does not uniquely identify each record.

### 8.6 Handling Exact Duplicate Records

#### Question

Should completely duplicated records be removed from the `geolocation` table?


In [76]:
geolocation_df.duplicated().sum()

np.int64(261831)

#### Insight

The geolocation table contains 261,831 exact duplicate rows. Removing these duplicates would reduce the table from 1,000,163 to 738,332 rows without losing unique geographic information.

### 8.7 Primary Key Re-validation After Deduplication

#### Question
Can a column or combination of columns uniquely identify each geographic record after removing exact duplicates?

-`geolocation_zip_code_prefix`

In [78]:
geolocation_df.drop_duplicates().shape[0]

738332

In [82]:
geolocation_df.drop_duplicates()["geolocation_zip_code_prefix"].nunique()

19015

It is not a PK.

- `geolocation_zip_code_prefix` + `geolocation_lat` + `geolocation_lng`

In [85]:
geolocation_df.drop_duplicates()[
    ["geolocation_zip_code_prefix",
     "geolocation_lat",
     "geolocation_lng"]
].drop_duplicates().shape[0]

720154

It is not a PK.
- All columns

In [89]:
geolocation_df.drop_duplicates()[
    ["geolocation_zip_code_prefix",
     "geolocation_lat",
     "geolocation_lng",
     "geolocation_city",
     "geolocation_state"]
].drop_duplicates().shape[0]

738332

#### Insight

After removing exact duplicates, the complete set of geographic attributes uniquely identifies each remaining row. However, these attributes are not necessarily a suitable primary key.